draft-jennings-moq-secure-objects.md

```diff
@@ -201,326 +190,248 @@ name but different data.
+|                                                       |   |
+|                                                       |   |
+|                                                       |   |
+|                                                       |   |
+|                   Encrypted Data                      |   |
+|                                                       |   +-- MOQT Object
+|                                                       |   |   Payload
+|                                                       |   |
+|                                                       |   |
++-------------------------------------------------------+   |
+|                 Authentication Tag                    |   |
++-------------------------------------------------------+ -'
+~~~
+{: #fig-sframe-ciphertext title="An SFrame, as it relates to MOQT" }
+
+For protect/unprotect of MOQT objects, the KID, CTR and metadata fields are
+computed from the MOQT object:
+
+* `KID = 0`
+* `CTR = (group_id << 32) | (object_id & 0xFFFFFFFF)`
+* `metadata = subscribe_id || ... || object_status` (i.e., the entirety of the
+  message aside from the payload)
+
+In order for this mapping to be one-to-one (which is required to avoid nonce
+reuse), both the Group ID and Object ID fields in the MOQT object MUST be less
+than 2<sup>32</sup>.
+
+> **NOTE:** If this limitation is uncomfortable, it can be mitigated in a few
+> ways, for example:
+> * having a key management scheme that allows for a track to be re-keyed
+> * Allowing the Group ID and Object ID subfields of the CTR value to have
+>   different sizes, so that for example an audio stream could have 0 object ID
+>   bits (thus requiring the object ID to always be zero) and 64 bits of group
+>   ID.
+
+## Protect
+
+To construct an MOQT secure object from an unprotected MOQT object, we perform
+an SFrame encryption and then throw away the SFrame header (since it can be
+reconstructed by the other side).
+
+1. Select the appropriate SFrame context for the track (see
+   {{setup-assumptions}}).
+
+2. Compute the KID, CTR, and metadata values for the object (see
+   {{mapping-between-sframe-and-moqt}}).
+
+3. Perform an SFrame encryption with the computed KID, CTR, and metadata values,
+   and the plaintext parameter set to the payload of the MOQT object (as
+   described in {{Section 4.4.3 of I-D.ietf-sframe-enc}}), returning an SFrame
+   ciphertext object.
+
+4. Construct an MOQT secure object of the same type as the input object, with
+   the following context:
+    * Set the payload to the contents of the "encrypted data" and
+      "authentication tag" portion of the SFrame ciphertext.
+    * Set the other fields of the message to the corresponding fields from the
+      input message.
+
+Note that the offset of the required bytes in the SFrame ciphertext can be
+computed by parsing the "config byte" which is the first byte of the SFrame
+ciphertext.
+
+~~~ pseudocode
```

Comment:
I hate this sort of pseudocode in specs. It can never decide if it is normative or not and duplicates the text in the draft.  I would rather say the normative stuff just once in clear english that explains what needs to happen not W3C style here is how you must code with the assumption the whole world is single threaded run to completion.

Reply:
It's a summary of the text.  I find it easier to read.  De gustibus.